cpp/include/cudf/join.hpp

```diff
@@ -393,128 +510,75 @@ class hash_join {
-   * the `left` portion and the corresponding columns in the `right` portion are omitted. To better
-   * align with the legacy `cudf::inner_join()` API, a `common_columns_output_side` parameter is
-   * introduced to specify whether "common" columns should go in `probe` or `build` portion.
-   *
-   * More details please @see cudf::inner_join().
+   * Returns the row indices that can be used to construct the result of performing
+   * an inner join between two tables. @see cudf::inner_join().
    *
    * @param probe The probe table, from which the tuples are probed.
-   * @param probe_on The column indices from `probe` to join on.
-   * @param columns_in_common is a vector of pairs of column indices into
-   * `probe` and `build`, respectively, that are "in common". For "common"
-   * columns, only a single output column will be produced, which is gathered
-   * from `probe_on` columns or `build_on` columns if `probe_output_side` is LEFT or RIGHT.
-   * Else, for every column in `probe_on` and `build_on`,
-   * an output column will be produced. For each of these pairs (P, B), P
-   * should exist in `probe_on` and B should exist in `build_on`.
-   * @param common_columns_output_side @see `common_columns_output_side`.
    * @param compare_nulls Controls whether null join-key values should match or not.
    * @param mr Device memory resource used to allocate the returned table and columns' device
    * memory.
    * @param stream CUDA stream used for device memory operations and kernel launches
    *
-   * @return Table pair of (`probe`, `build`) of joining both tables on the columns
-   * specified by `probe_on` and `build_on`. The resulting table pair will be joined columns of
-   * (`probe(including common columns)`, `build(excluding common columns)`) if
-   * `common_columns_output_side` is `PROBE`, or (`probe(excluding common columns)`,
-   * `build(including common columns)`) if `common_columns_output_side` is `BUILD`.
+   * @return A pair of columns [`left_indices`, `right_indices`] that can be used to construct
+   * the result of performing an inner join between two tables with `build` and `probe`
+   * as the the join keys .
    */
-  std::pair<std::unique_ptr<cudf::table>, std::unique_ptr<cudf::table>> inner_join(
-    cudf::table_view const& probe,
-    std::vector<size_type> const& probe_on,
-    std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
-    common_columns_output_side common_columns_output_side = common_columns_output_side::PROBE,
-    null_equality compare_nulls                           = null_equality::EQUAL,
-    rmm::cuda_stream_view stream                          = rmm::cuda_stream_default,
-    rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource()) const;
+  std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+            std::unique_ptr<rmm::device_uvector<size_type>>>
+  inner_join(cudf::table_view const& probe,
```

Comment:
We need some documentation on the hash_join object that any dictionary columns in build/probe are expected to already be matched with a reference to the API that can be used to match them.